crates/crypto_api/src/crypto_system/mod.rs

```diff
@@ -48,11 +54,23 @@ pub trait CryptoSystem: Sync {
         salt: &Box<dyn Buffer>,
     ) -> CryptoResult<()>;
 
-    /// compute a sha256 hash for `data`, storing it in `hash`
-    fn hash_sha256(&self, hash: &mut Box<dyn Buffer>, data: &Box<dyn Buffer>) -> CryptoResult<()>;
+    /// bytelength of key derivation context
+    fn kdf_context_bytes(&self) -> usize;
 
-    /// compute a sha512 hash for `data`, storing it in `hash`
-    fn hash_sha512(&self, hash: &mut Box<dyn Buffer>, data: &Box<dyn Buffer>) -> CryptoResult<()>;
+    /// minimum bytelength of key derivation buffers
+    fn kdf_min_bytes(&self) -> usize;
+
+    /// maximum bytelength of key derivation buffers
+    fn kdf_max_bytes(&self) -> usize;
+
+    /// derive a new deterministic key based of index, context, and parent
+    fn kdf(
+        &self,
+        out_buffer: &mut Box<dyn Buffer>,
+        index: u64,
+        context: &Box<dyn Buffer>,
+        parent: &Box<dyn Buffer>,
+    ) -> CryptoResult<()>;
```

Comment:
It seems a little too easy to use the wrong buffer here in the argument list, maybe in a future iteration we switch to a builder pattern for this?

Reply:
named param macros!